Feature/#268 solution for storing large intermediate results in db

yannick-vinkesteijn opened the pull request:
This pull request introduces significant changes to enhance data streaming, improve dependency management, and refactor the codebase for better maintainability and functionality. The most important updates include implementing streaming for large inputs and results, updating Docker configurations, and simplifying imports and encryption methods.

### Enhancements to Data Streaming:
* Updated the `wait_for_results` and `create` methods in `vantage6-algorithm-tools` and `vantage6-client` to stream large inputs and results in chunks, improving performance and reliability for large data transfers. [[1]](diffhunk://#diff-cf417e1eed464b9fb5d06551c300ab7f84e8059662780518833b033cd93a0a30L145-R172) [[2]](diffhunk://#diff-69a6539ff08614cba317f57bddb5f0ff560da1a9adb1e9f9d5cf6ff698b34ddaL237-R255) [[3]](diffhunk://#diff-cf417e1eed464b9fb5d06551c300ab7f84e8059662780518833b033cd93a0a30L378-R438) [[4]](diffhunk://#diff-69a6539ff08614cba317f57bddb5f0ff560da1a9adb1e9f9d5cf6ff698b34ddaL1873-R1922) [[5]](diffhunk://#diff-350128e5674e17ec3381b6ed104ecbd380e31f4086514a7a8d120215a6b40906L187-R233) [[6]](diffhunk://#diff-350128e5674e17ec3381b6ed104ecbd380e31f4086514a7a8d120215a6b40906L238-R310)
* Added logic to handle UUID-based input and result references, ensuring compatibility with the new streaming approach. [[1]](diffhunk://#diff-350128e5674e17ec3381b6ed104ecbd380e31f4086514a7a8d120215a6b40906L187-R233) [[2]](diffhunk://#diff-350128e5674e17ec3381b6ed104ecbd380e31f4086514a7a8d120215a6b40906L238-R310)

### Updates to Docker Configurations:
* Enhanced the `uwsgi` command in `docker-compose.yml` with additional parameters for better HTTP handling, such as chunked input and keep-alive support.
* Added dependency installation from `requirements.txt` in `docker/algorithm-base.Dockerfile` to ensure consistent package resolution.

### Codebase Refactoring and Simplification:
* Replaced `json_lib` with `json` across multiple files for consistency and simplicity. [[1]](diffhunk://#diff-cf417e1eed464b9fb5d06551c300ab7f84e8059662780518833b033cd93a0a30L4-R4) [[2]](diffhunk://#diff-27b199273137959513970489cd8e76fce35dd7c1317cc974917aac29334a9dc5L4-R4) [[3]](diffhunk://#diff-27b199273137959513970489cd8e76fce35dd7c1317cc974917aac29334a9dc5L266-R266) [[4]](diffhunk://#diff-cf417e1eed464b9fb5d06551c300ab7f84e8059662780518833b033cd93a0a30L268-R288) [[5]](diffhunk://#diff-cf417e1eed464b9fb5d06551c300ab7f84e8059662780518833b033cd93a0a30L307-R327)
* Removed unused `Node` property from the `AlgorithmClient` class, streamlining its initialization.

### Encryption and Decryption Improvements:
* Updated the `encrypt_bytes_to_str` and `decrypt_str_to_bytes` methods in `vantage6-common` to handle byte-based encryption more effectively.

These changes collectively improve the system's scalability, maintainability, and performance, particularly for handling large data payloads in distributed environments.

tim-hendriks committed:
Add early version of azure data streaming

tim-hendriks committed:
Fix issues with node and algorithm client

tim-hendriks committed:
Remove erroneous addition of self.node in algorithm client

tim-hendriks committed:
Allow for azure service principal account connection

tim-hendriks committed:
Add streaming functionality to proxy, altering result retrieval

tim-hendriks committed:
Changes for input blob storage testing, some of these should be reverted

tim-hendriks committed:
Make changes to fix input streaming issues

tim-hendriks committed:
Add encryption logic for large results

tim-hendriks committed:
Remove logging of actual data, fix output result structure in wait_for_results

tim-hendriks committed:
Remove confusing import renames, touch-reload

yannick-vinkesteijn commented:
**To build vantage6 node/server container**
- new version numbers
- Dockerfile gebruiken -> via makefile via standaard route
- pip packages kan gemaakt worden, hoeft niet via publishen
- Controleren of pluginML zonder json ook goed werkt
- vooral encryption stuk laten checken door v6 team
- andere formats controleren of die ook door encryption flow komen -> bijv safetensor, parquet

MarlouvS committed:
Merge branch 'main' of https://github.com/DutchHospitalData/vantage6 into feature/#268-SolutionForStoringLargeIntermediateResultsInDb

MarlouvS committed:
Adding logging and errorhandling to new code of feature #268

MarlouvS committed:
Adding tests for client

MarlouvS commented on vantage6-common/vantage6/common/encryption.py:
```diff
@@ -294,77 +371,174 @@ def encrypt_bytes_to_str(self, data: bytes, pubkey_base64s: str) -> str:
+        if final_chunk:
+            yield final_chunk
+
+    def encrypt_stream(self, stream, pubkey_base64s: str, chunk_size=8192):
```
Encrypt_stream en decrypt_stream worden 2 maal aangemaakt (in CryptorBase en RSACryptor), is het de bedoeling dat deze functie overschreven wordt?

MarlouvS commented on vantage6-server/vantage6/server/service/azure_storage_service.py:
```diff
@@ -0,0 +1,55 @@
+from vantage6.common import logger_name
+
+module_name = logger_name(__name__)
+log = logging.getLogger(module_name)
```
Waarom verschilt de manier van logging hier van eerder in dit pull request? Zoals bij vantage6-common/vantage6/common/client/client_base.py

MarlouvS commented on vantage6-server/vantage6/server/resource/result.py:
```diff
@@ -0,0 +1,191 @@
+        result, self._buffer = self._buffer[:size], self._buffer[size:]
+        return result
+  
+class ResultStreams(ResultStreamBase):
```
ResultStream en ResultStreams kunnen waarschijnlijk samengevoegd worden tot 1 class met een get en post methode

MarlouvS committed:
Made changes in the Resultstream to combine the Get and Post method. Changed resultstreamer from uwsgi to werkzeug.wsgi

tim-hendriks committed:
Use new wait_for_task_completion in algorithm client

MarlouvS committed:
Merge branch 'main' of https://github.com/vantage6/vantage6 into #268_code_placement_improvement

tim-hendriks committed:
Allow for azurite testing using server config yaml with connection string

tim-hendriks committed:
Add automatic importing of entities for integration testing

tim-hendriks committed:
Change test entities to comply with test configurations

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into #268_code_placement_improvement

tim-hendriks committed:
Add test algorithm to integration tests folder

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into #268_code_placement_improvement

MarlouvS committed:
changed spot of resultstream post + fix small bug

MarlouvS committed:
Remove broken tests for which code changed, still need to be added back

MarlouvS committed:
remove azure test for now

MarlouvS committed:
removal of duplicate code

MarlouvS committed:
Reverting removal pubkey_base64s variable

tim-hendriks committed:
Merge pull request #5 from DutchHospitalData/#268_code_placement_improvement

#268 code placement improvement

tim-hendriks committed:
Add documentation for large results

tim-hendriks committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of git+ssh://github.com/mdw-nl/vantage6 into feature/#268-SolutionForStoringLargeIntermediateResultsInDb

tim-hendriks committed:
Fix len() error for streams in azure service

MarlouvS committed:
renaming resultstream and wait_for_results

MarlouvS committed:
Added documentation to methodes

MarlouvS committed:
Merge branch 'main' of https://github.com/vantage6/vantage6 into #268_proper_naming

tim-hendriks committed:
Merge pull request #6 from DutchHospitalData/#268_proper_naming

#268 proper naming

tim-hendriks committed:
Refactor and optimize code across multiple files

- Removed unnecessary buffer size parameter from uwsgi commands in server.sh, start.py, and client_base.py.
- Simplified the Docker Compose configuration by removing redundant buffer size settings.
- Enhanced the server entrypoint script by cleaning up the uwsgi command.
- Updated test-entities.yaml to remove IDs for collaborations and organizations, this does not work.
- Cleaned up requirements.txt by removing aiohttp dependency.
- Added new methods for uploading and downloading run data in client_base to deduplicate existing functionality across clients and node.
- Introduced input validation checks in task.py to ensure UUIDs are used for large result stores.
- Refactored the setup of large result store in server/__init__.py for better clarity and maintainability.
- Added a new endpoint for checking blob store status in result.py. This is used in the UserClient to ascertain whether the first input needs to be stored in blob store or not.
- Improved error handling and logging throughout the codebase.

tim-hendriks committed:
Add checks for backwards compatibility

- Changed large result store type from "azure_blob_storage" to "azure" in configuration files.
- Updated integration tests to reflect changes in organization handling.
- Introduced DataStorageUsed enum to manage storage types across the application.
- Implemented blob stream resource for handling large result uploads and downloads.
- Enhanced run model to include data_storage_used attribute.
- Updated API schemas and frontend components to accommodate new data storage information.

MarlouvS committed:
Delete blobs from azure

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into feature/deletion_of_blobs

tim-hendriks committed:
Fix backwards compatibility

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into feature/deletion_of_blobs

tim-hendriks committed:
Remove integration test folder

tim-hendriks committed:
Use result instead of generic request when retrieving results

tim-hendriks committed:
Fix duplicate result retrieval

tim-hendriks committed:
Small refactor of algorithm task create

tim-hendriks committed:
Remove print statements from clientbase

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into feature/deletion_of_blobs

tim-hendriks committed:
Fix duplicate decrypting of results

tim-hendriks committed:
Remove print statements and duplicate return

tim-hendriks committed:
Add alerts for blob inputs and results to frontend

tim-hendriks committed:
Fix existing unit tests and small release notes correction

tim-hendriks committed:
Fix unit tests

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into feature/deletion_of_blobs

MarlouvS committed:
Add deletion of blobs to task deletion

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into feature/deletion_of_blobs

tim-hendriks committed:
Add comments, docstrings and change logging levels

tim-hendriks committed:
Add initial developer documentation

MarlouvS committed:
fixed delection of blob when task is deleted

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into feature/deletion_of_blobs

MarlouvS committed:
remove integration_tests

tim-hendriks committed:
Fix addition of result layer in from_task

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into feature/deletion_of_blobs

MarlouvS committed:
Fix tests Clients

MarlouvS committed:
Add blobstorage cleanup to cleanup.py

MarlouvS committed:
Changes because of review points

MarlouvS committed:
Get datastorage used from run instead of check_if_blob_store_enabled()

MarlouvS committed:
Adding checks to cleanup

tim-hendriks committed:
Merge pull request #7 from DutchHospitalData/feature/deletion_of_blobs

Feature/deletion of blobs

tim-hendriks committed:
Reformat whitespace

tim-hendriks committed:
Remove addition of version tag in algorithm base

tim-hendriks committed:
Reduce codacy issues

tim-hendriks committed:
Update docs/server/optional.rst

Co-authored-by: Bart van Beusekom <bartvanbeusekom@gmail.com>

tim-hendriks committed:
Update docs/server/yaml/server_config.yaml

Co-authored-by: Bart van Beusekom <bartvanbeusekom@gmail.com>

tim-hendriks committed:
Remove uwsgi from requirements as it's already installed from source. Add it to setup.py

tim-hendriks committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of git+ssh://github.com/mdw-nl/vantage6 into feature/#268-SolutionForStoringLargeIntermediateResultsInDb

tim-hendriks committed:
Revert wait_for_results renaming

tim-hendriks committed:
Revert docstring changes

tim-hendriks committed:
Change to string encoding global constant instead of hardcoded utf-8

tim-hendriks committed:
Remove print from task deletion

tim-hendriks committed:
Use 'results' rather than 'runs' for response from result endpoint

yannick-vinkesteijn committed:
algo tools init test review fix

tim-hendriks committed:
Fix blob store documentation issues

tim-hendriks committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of git+ssh://github.com/mdw-nl/vantage6 into feature/#268-SolutionForStoringLargeIntermediateResultsInDb

tim-hendriks committed:
Add mutual exclusiveness clarification to connection string/service principal

tim-hendriks committed:
Make clientbase methods private, fix organization input empty dict bug in algorithm client, add import after monkey patch, add docstring fixes

tim-hendriks committed:
Change log level of public key printing

tim-hendriks committed:
Remove algorithm tools globals

tim-hendriks committed:
Use self.headers in client base

tim-hendriks committed:
Add exception if unable to upload run data

tim-hendriks committed:
Improve is_uuid util

tim-hendriks committed:
Update azure service; update docstrings and init parameters

yannick-vinkesteijn committed:
fixed patch comment for init and client tests

yannick-vinkesteijn committed:
fixed patch comment for client tests

tim-hendriks committed:
Fix tests

tim-hendriks committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of git+ssh://github.com/mdw-nl/vantage6 into feature/#268-SolutionForStoringLargeIntermediateResultsInDb

tim-hendriks committed:
Remove enums for different blob storages

tim-hendriks committed:
Enforce permissions in blobstream

tim-hendriks committed:
Add after delete for run to blob service

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into #268_delete_orphan_blobs

tim-hendriks committed:
Fix session bug

tim-hendriks committed:
Remove blob_store_enabled variable only used once

tim-hendriks committed:
Remove todo task manager

tim-hendriks committed:
Add whitespace to docstring proxy server

tim-hendriks committed:
Fix import order server init

tim-hendriks committed:
Change config name server to avoid confusion

tim-hendriks committed:
Remove checking for enums from server output_json

tim-hendriks committed:
Fix incorrect status blobstream endpoint

MarlouvS committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of https://github.com/mdw-nl/vantage6 into #268_delete_orphan_blobs

tim-hendriks committed:
Update non descriptive return message in blobstream

tim-hendriks committed:
Add blobstore to existing list when loading resources

tim-hendriks committed:
Add storage adapter to injected services

tim-hendriks committed:
Fix uwsgi related error

tim-hendriks committed:
Fix uwsgi related error

tim-hendriks committed:
make PR test related changes

tim-hendriks committed:
Make download method private, fix is_uuid method

MarlouvS committed:
move delete blob to delete task function

tim-hendriks committed:
Remove docker clean up related code for another PR

tim-hendriks committed:
Remove delete blob functionality from endpoint

tim-hendriks committed:
Fix tests, add setup to test__init

tim-hendriks committed:
Remove large result store from cleanup

tim-hendriks committed:
Remove unnessecary constants from testresources

tim-hendriks committed:
Fix whitespace/formatting

tim-hendriks committed:
Merge remote-tracking branch 'dhd/#268_delete_orphan_blobs' into feature/#268-SolutionForStoringLargeIntermediateResultsInDb

tim-hendriks committed:
Remove test_task

tim-hendriks committed:
Fix test issues, test delete blobs

tim-hendriks committed:
Fix comments on outdated files

tim-hendriks committed:
Refactor encryption and fix client_base mixin

tim-hendriks committed:
Fix formatting

tim-hendriks committed:
Update vantage6-server/vantage6/server/resource/task.py

Co-authored-by: Bart van Beusekom <bartvanbeusekom@gmail.com>

tim-hendriks committed:
Update docs/features/inter-component/blob_storage.rst

Co-authored-by: Bart van Beusekom <bartvanbeusekom@gmail.com>

tim-hendriks committed:
Update docs/features/inter-component/blob_storage.rst

Co-authored-by: Bart van Beusekom <bartvanbeusekom@gmail.com>

tim-hendriks committed:
Make various smaller PR changes

tim-hendriks committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of git+ssh://github.com/mdw-nl/vantage6 into feature/#268-SolutionForStoringLargeIntermediateResultsInDb

tim-hendriks committed:
Add error when unexpected data type for decrypt.

tim-hendriks committed:
Change azure service to take in config, add try catch to cleanup

tim-hendriks committed:
Add bytes return type to encrypt_bytes_to_str

tim-hendriks committed:
Slightly modify comment in utest

tim-hendriks committed:
Update vantage6-common/vantage6/common/client/client_base.py

Co-authored-by: Bart van Beusekom <bartvanbeusekom@gmail.com>

tim-hendriks committed:
Update vantage6-common/vantage6/common/encryption.py

Co-authored-by: Bart van Beusekom <bartvanbeusekom@gmail.com>

tim-hendriks committed:
Add minor PR changes

tim-hendriks committed:
Merge branch 'feature/#268-SolutionForStoringLargeIntermediateResultsInDb' of git+ssh://github.com/mdw-nl/vantage6 into feature/#268-SolutionForStoringLargeIntermediateResultsInDb

tim-hendriks committed:
Remove unused imports

tim-hendriks committed:
Fix unit test after parameter change, add permissions to post

tim-hendriks committed:
Update vantage6-server/vantage6/server/resource/blobstream.py

Co-authored-by: Bart van Beusekom <bartvanbeusekom@gmail.com>

tim-hendriks committed:
Revert to old uwsgi openssl paths in dockerfile for now

tim-hendriks committed:
Revert whitespacing in dockerfile
